src/vdaf/idpf.rs

```diff
@@ -0,0 +1,1620 @@
+                            for (dest, src) in array.iter_mut().zip(other_array.iter()) {
+                                *dest += *src;
+                            }
+                        }
+                        IdpfOutputShare::Leaf(_) => {
+                            return Err(VdafError::Uncategorized(
+                                "mismatched field types".to_string(),
+                            ))
+                        }
+                    }
+                }
+                Ok(IdpfOutput::Inner(array))
+            }
+            IdpfOutputShare::Leaf(self_array) => {
+                let mut array = *self_array;
+                for other in others {
+                    match other {
+                        IdpfOutputShare::Inner(_) => {
+                            return Err(VdafError::Uncategorized(
+                                "mismatched field types".to_string(),
+                            ))
+                        }
+                        IdpfOutputShare::Leaf(other_array) => {
+                            for (dest, src) in array.iter_mut().zip(other_array.iter()) {
+                                *dest += *src;
+                            }
+                        }
+                    }
+                }
+                Ok(IdpfOutput::Leaf(array))
+            }
+        }
+    }
+}
+
+/// The IDPF scheme used in Poplar1.
+#[derive(Debug, Clone)]
+pub struct IdpfPoplar<P, const L: usize, const OUT_LEN: usize> {
+    bits: usize,
+    _phantom: PhantomData<(P, [(); OUT_LEN])>,
+}
+
+impl<P, const L: usize, const OUT_LEN: usize> IdpfPoplar<P, L, OUT_LEN>
+where
+    P: Prg<L>,
+{
+    /// Construct this IDPF by providing the size in bits of inputs it will process.
+    pub fn new(bits: usize) -> Result<Self, VdafError> {
+        if bits == 0 {
+            return Err(VdafError::Uncategorized(
+                "invalid number of bits: 0".to_string(),
+            ));
+        }
+        Ok(IdpfPoplar {
+            bits,
+            _phantom: PhantomData,
+        })
+    }
+
+    /// An auxiliary function that acts as a pseudorandom generator, extending one seed into two
+    /// seeds plus some additional bits.
+    fn extend(&self, seed: &Seed<L>) -> ([Seed<L>; 2], [bool; 2]) {
+        let mut prg = P::init(&seed.0);
+        prg.update(VERSION);
+        prg.update(b" idpf poplar extend");
+        let mut seed_stream = prg.into_seed_stream();
+
+        let mut seeds = [Seed([0u8; L]), Seed([0u8; L])];
+        seed_stream.fill(&mut seeds[0].0);
+        seed_stream.fill(&mut seeds[1].0);
+
+        let mut byte = [0u8];
+        seed_stream.fill(&mut byte);
+        let control_bits = [(byte[0] & 1) != 0, ((byte[0] >> 1) & 1) != 0];
+
+        (seeds, control_bits)
+    }
+
+    /// An auxiliary function that acts as a pseudorandom generator, returning field elements.
+    fn convert<F>(&self, seed: &Seed<L>) -> (Seed<L>, [F; OUT_LEN])
```

Comment:
Suggestion: It might be easier to oparate on `[u8; L]` directly rather than `Seed<L>`.